Extended insert format to search-replace SQL export

villevuor opened the pull request:
When exporting SQL dump with search-replace command, every row of DB is inserted on separate `INSERT INTO`query. This is quite heavy, especially on large databases. (For example check issue #2634.)

Solution is to use "extended insert format", which basicly adds all the insertions to same query.

danielbachhuber commented:
This might have a negative performance impact on sites with millions of rows. Can you do some performance profiling, and share the results?

danielbachhuber commented:
> This might have a negative performance impact on sites with millions of rows.

Meaning, you're essentially reading the entire table into memory before writing it out.

Instead, I think we'd need to do chunked insert statements if we want to break them apart. Sequel Pro does this based on data size:

![image](https://cloud.githubusercontent.com/assets/36432/15250025/8c8de118-18d7-11e6-8ad9-bea603802627.png)

We could do based on data size (`--export_chunk_size=<size-kb>`), or based on # of rows.

Lastly, you'll need to include some updated functional tests reflecting these changes. If you introduced a chunk size flag, then it would be easier to test, as you could set it to a small chunk size in the test and only need to generate a limited size of mock data.

villevuor commented:
Thanks for comments!

I just committed a feature that breaks the the insert statement into chunks. After the weekend I'll check out some performance data and maybe add option to specify the chunk size.

danielbachhuber commented on php/commands/search-replace.php:
```diff
@@ -328,20 +331,48 @@ private function php_handle_col( $col, $primary_keys, $table, $old, $new ) {
-		$sql = $wpdb->prepare( $sql, array_values( $row_fields ) );
-		fwrite( $this->export_handle, $sql );
+
+		if(!empty($rows)) {
```
For better readability, you can bail early here:

```
if ( empty( $rows ) ) {
    return;
}
```

danielbachhuber commented on features/search-replace-export.feature:
```diff
@@ -11,7 +11,11 @@ Feature: Search / replace with file export
+      """
+    And STDOUT should contain:
+      """
+      ('1', 'siteurl', 'http://example.net', 'yes')
```
We need a test that multiple chunks get created, and that those multiple chunks are appropriately inserted into the database.

villevuor commented:
Thanks again. I fixed the parts you pointed. (I also found a mistake in my previous commits that slowed export a lot, but it’s now fixed.)

I did some performance testing with exporting. I basicly ran `wp search-replace 'http://example.com' 'http://example2.com' --export=dump.sql` and measured time with `time` command. Import times were tested on command line running `mysql -u user -p dbname < dump.sql --debug-info`.

**Search-replace export with WP CLI 0.23.1**

Posts in WP install | 200 | 3000 | 35 000
------------ | ------------- | ------------- | -------------
Export file size | 191 kb | 2 863 kb | 28 287 kb
Export time | 1.655s | 1.884s | 11.109s
Import time (locally) | 1.1s | 5.3s | 79.8s
Import time (VPS) | 2.3s | 11.3s | 146.1s

**Search-replace export with this branch**

Posts in WP install | 200 | 3000 | 35 000
------------ | ------------- | ------------- | -------------
Export file size | 89 kb | 893 kb | 10 686 kb
Export time | 1.284s | 1.770s | 9.065s
Import time (locally) | 1.1s | 1.2s | 2.4s
Import time (VPS) | 1.2s | 1.8s | 4.2s

I didn’t test with *millions* of rows, but this already seems good. Export was tested on my Macbook Pro (8 gt i5). Import was tested on my local machine and also on cheapest VPS from Scaleway.

villevuor commented:
I think next I could implement the chunk size flag and find out what is the best default value. After that I think this starts to look good?

danielbachhuber commented:
> I didn’t test with millions of rows, but this already seems good.

Yep, this does look like a substantial improvement. Thanks for sticking with it :)

> I think next I could implement the chunk size flag and find out what is the best default value. After that I think this starts to look good?

I'm not actually convinced we need a chunk size option. Splitting based on number of rows seems like it could be a good first version. What do you think?

Also, can you fix the failing tests on 3.7?

villevuor commented:
I added `export_insert_size` flag that defines number of rows in single `INSERT` query. It would be good to check that out, I'm not sure did I do the flag validation the best way. I think number of rows is fair enough at this points.

I'll check out the failing tests tomorrow.

danielbachhuber commented:
> I'll check out the failing tests tomorrow.

If you merge master, that will address the failing `trunk` build

villevuor commented:
Have you experienced this before?

`No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.`

https://travis-ci.org/wp-cli/wp-cli/jobs/130767916

danielbachhuber commented:
> Have you experienced this before?

Given the context, it appears your code is incompatible in some form with PHP 5.3

villevuor commented:
>Given the context, it appears your code is incompatible in some form with PHP 5.3.

New merge solved this.

Everything done now? :)

danielbachhuber commented on php/commands/search-replace.php:
```diff
@@ -61,6 +62,9 @@ class Search_Replace_Command extends WP_CLI_Command {
 	 * : Write transformed data as SQL file instead of saving replacements to
 	 * the database. If <file> is not supplied, will output to STDOUT.
 	 *
+	 * [--export_insert_size[=<rows>]]
```
This should be `[--export_insert_size=<rows>]`

Note that it shouldn't ever be used as a flag.

danielbachhuber commented on php/commands/search-replace.php:
```diff
@@ -61,6 +62,9 @@ class Search_Replace_Command extends WP_CLI_Command {
 	 * the database. If <file> is not supplied, will output to STDOUT.
 	 *
+	 * [--export_insert_size[=<rows>]]
+	 * : Define number of rows in single INSERT statement when doing SQL export.
```
Can you add a short explanation as to why you might use or change this value?

Also, to specify the default value in the document, you can use inline YAML doc:

```
---
default: 50
---
```

danielbachhuber commented on php/commands/search-replace.php:
```diff
@@ -328,20 +341,51 @@ private function php_handle_col( $col, $primary_keys, $table, $old, $new ) {
+		$count = count( $rows );
+		$export_insert_size = $this->export_insert_size;
+
+		foreach($rows as $row_fields) {
```
Is there a better way to express this `foreach()`?

villevuor replied:
What you mean by this?

danielbachhuber replied:
```
if( ( $index % $export_insert_size == 0 && $index > 0 ) || $index == $count ) {
```

Meaning this is a complex conditional and difficult to review / understand the logic of. Can it be written more simply?

danielbachhuber commented on php/commands/search-replace.php:
```diff
@@ -124,6 +128,12 @@ public function __invoke( $args, $assoc_args ) {
 					WP_CLI::error( sprintf( 'Unable to open "%s" for writing.', $assoc_args['export'] ) );
 				}
 			}
+			$export_insert_size = \WP_CLI\Utils\get_flag_value( $assoc_args, 'export_insert_size' );
```
Once you specify the default value per my comment above, you can just access `$assoc_args['export_insert_size']` directly.

danielbachhuber commented:
> Everything done now? :)

Left some comments.

villevuor commented:
>Meaning this is a complex conditional and difficult to review / understand the logic of. Can it be written more simply?

I think it is quite straightforward now. Added some comments though to clarify the logic.

danielbachhuber commented:
> I think it is quite straightforward now. Added some comments though to clarify the logic.

Looks good 👍 Thanks for your time on this!
